### ISSUE TITLE
Dict values not inferred correctly with .get() calls

### ISSUE THREAD

**Issue by davidbauer99** [python/mypy#6258]
The function below results in an unexpected return type error. `test.py:6: error: Incompatible return value type (got "Optional[int]", expected "int")`

```python
from typing import Dict


def results_in_error(data: Dict[str, int], key: str) -> int:
    if key in data:
        return data.get(key)
    return 0


def works_as_expected(data: Dict[str, int], key: str) -> int:
    val = data.get(key)
    if val:
        return val
    return 0
```
Replacing `if key in data:` with `if data.get(key):` will result in an identical error. 

My mypy.ini file is the following:
```
[mypy]
ignore_missing_imports=True
python_version = 3.6
strict_optional=True
check_untyped_defs=True
disallow_untyped_calls=True
disallow_untyped_defs=True
follow_imports=normal
follow_imports_for_stubs=True
```

I believe this is a false positive as the value will not be `None` due to the dict's typing.

**Comment by ilevkivskyi** [comment]
From typing point of view there is nothing wrong here. For comparison:
```python
def test() -> int:
    if 2 + 2 > 4:
        return 'no way'
```
will also fail. It is just not possible to special-case all possible patterns. In particular, your original example is actually an anti-pattern. Instead, you should write:
```python
def better_way(data: Dict[str, int], key: str) -> int:
    if key in data:
        return data[key]
    return 0
```
Or even better:
```python
def even_better_way(data: Dict[str, int], key: str) -> int:
    return data.get(key, 0)
```

**Comment by unformatt** [comment]
I don't understand this explanation. Showing a "better" way to write the code doesn't explain why mypy doesn't think that `data.get(key)` would return an `int` given that `data` is typed as `Dict[str, int]`.

**Comment by cxong** [comment]
`get(key)` returns `None` if the key is not found

**Comment by sshishov** [comment]
The explanation is simple: when it is executing the line `return data.get(key)`, the signature of this functions is returning `Optional[int]` because if the key does not exist, it will return `None`.

It cannot know by this time that the key does exist because otherwise it should store some kind of state machine to know that we have already checked that the key in the dict.

As suggested, we should be more specific and write the code which cannot be interpreted differently. If you expect on this line where you return, that the value DOES EXIST in the dictionary, just return it using `data[key]` which cannot return None and therefore will not return `Optional`

**Comment by caliraftdude** [comment]
I get the difficulty here in trying to check data driven code that might be ambiguous, but the suggestion of using data[key] as an alternative seems more like a suggestion on how to avoid a "mypy" error as opposed to avoiding the program from crashing because of a non-existent key.  The exact reason for the get() function is to short circuit a more complex "look for key, validate it exists, then try and retrieve it" paradigm by just returning the value or None.

I would have to disagree with this, and a better option would be to suppress this specific error if indeed the code is sound but too ambiguous to analyze.

**Comment by hauntsaninja** [comment]
Also note that if the default value is expensive to construct, you may want something like `data.get(key) or expensive_object()` or if your values can be false-y then `val if (val := data.get(key)) is not None else expensive_object()`
